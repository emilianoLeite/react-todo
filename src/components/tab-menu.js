import React, { Component } from "react";
import PropTypes from "prop-types";

export class TabMenu extends Component {
  constructor(props) {
    super(props);

    this.state = { currentTab: this.children.findIndex(child => child.props.active) };

    this.classesFor = this.classesFor.bind(this);
  }

  classesFor(menuItem) {
    if (this.state.currentTab === menuItem) {
      return "menu-item active";
    } else {
      return "menu-item";
    }
  }

  get children() {
    return React.Children.toArray(this.props.children);
  }

  render() {
    return (<div>
      <div className="filter-menu">
        {
          this.children.map((C, index) => {
            return {
              ...C, props: {
                ...C.props,
                onClick: () => { this.setState({ currentTab: index }); },
                className: this.classesFor(index)
              }
            };
          })
        }
      </div>
      {this.children[this.state.currentTab].props.render()}
    </div>);
  }
}


export class Tab extends Component {
  render() { return (
    <div className={this.props.className} onClick={this.props.onClick}>
      {this.props.name}
    </div>
  ); }
}

TabMenu.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired
};

Tab.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  render: PropTypes.func
};
