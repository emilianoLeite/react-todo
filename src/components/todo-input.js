import PropTypes from "prop-types";
import React from "react";
import { Todo } from "../models";
import { addTodo } from "../redux/actions";
import { connect } from "react-redux";

class TodoInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = { value: props.value };

    this.ENTER_KEY = 13;
    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.clearInput = this.clearInput.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleKeyPress(event){
    const pressedKey = (event.keyCode ? event.keyCode : event.which);

    if (pressedKey === this.ENTER_KEY) {
      this.props.addTodo(new Todo({ text: event.target.value }));
      this.clearInput();
    }
  }

  clearInput(){ this.setState({ value: "" }); }

  handleChange(event){
    this.setState({ value: event.target.value });
  }

  render() { return (
    <input
      type='text'
      placeholder='New todo...'
      autoFocus={true}
      value={this.state.value}
      onKeyPress={this.handleKeyPress}
      onChange={this.handleChange}
    />
  );}
}

TodoInput.propTypes = {
  value: PropTypes.string,
  onCreateTodo: PropTypes.func.isRequired,
  addTodo: PropTypes.func
};

TodoInput.defaultProps = {
  value: ""
};

export default connect(() => ({}), { addTodo })(TodoInput);
