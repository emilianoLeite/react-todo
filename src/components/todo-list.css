.filter-menu {
  display: flex;
  flex-direction: row;
  align-items: baseline;
}
.filter-menu .menu-item {
  padding: 3px;
  border: 1px solid black;
  border-left: none;
}

.filter-menu > :first-child.menu-item {
  border-left: 1px solid black;
}

.filter-menu .menu-item.active {
  border-bottom: none;
}
