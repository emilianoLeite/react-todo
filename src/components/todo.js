import PropTypes from "prop-types";
import React from "react";
import { TodoInput } from ".";
import { Todo as TodoModel } from "../models";

export default class Todo extends React.Component {
  constructor(props) {
    super(props);
    this.state = { editing: false };
    this.handleTodoUpdate = this.handleTodoUpdate.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  handleTodoUpdate(todoText) {
    const updatedTodo = new TodoModel({ ...this.props.todo, text: todoText });

    this.props.onUpdateTodo(updatedTodo);

    this.setState({ editing: false });
  }

  handleClick() {
    this.setState({ editing: true });
  }

  render() {
    if (this.state.editing) {
      return (
        <TodoInput
          onCreateTodo={this.handleTodoUpdate}
          value={this.props.todo.text}
        />
      );
    } else {
      return (
        <span
          onClick={this.handleClick}
          className="todo-text"
        >
          {this.props.todo.text}
        </span>
      );
    }
  }
}

Todo.propTypes = {
  todo: PropTypes.shape({
    text: PropTypes.string.isRequired,
    completed: PropTypes.bool
  }),
  onUpdateTodo: PropTypes.func.isRequired
};
