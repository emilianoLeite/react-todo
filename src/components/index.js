import { Tab, TabMenu } from "./tab-menu";
import Todo from "./todo";
import TodoInput from "./todo-input";
import TodoList from "./todo-list";

export { Tab, TabMenu, Todo, TodoInput, TodoList } ;

