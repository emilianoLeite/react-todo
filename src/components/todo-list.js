import "./todo-list.css";
import { Tab, TabMenu } from "./tab-menu";
import PropTypes from "prop-types";
import React from "react";
import Todo from "./todo";
import { Todo as TodoModel } from "../models";
import { connect } from "react-redux";
import { toggleTodo } from "../redux/actions";

class TodoList extends React.Component {
  constructor(props) {
    super(props);

    this.showAllTodos = this.showAllTodos.bind(this);
    this.showCompletedTodos = this.showCompletedTodos.bind(this);
    this.showIncompleteTodos = this.showIncompleteTodos.bind(this);
  }

  showAllTodos() {
    return this.showTodos(this.props.todos);
  }

  showCompletedTodos() {
    const todos = this.props.todos.filter(todo => todo.completed);
    return this.showTodos(todos);
  }

  showIncompleteTodos() {
    const todos = this.props.todos.filter(todo => !todo.completed);
    return this.showTodos(todos);
  }

  showTodos(todos) {
    return <ul className="todo-list">
      {
        todos.map((todo) => {
          return (
            <li
              className={todo.completed ? "completed" : ""}
              key={todo.id}
            >
              <input
                defaultChecked={todo.completed}
                type="checkbox"
                onClick={() => this.props.toggleTodo(todo)}
              />
              <Todo
                onUpdateTodo={this.props.onUpdateTodo}
                todo={todo}
              />
            </li>
          );
        })
      }
    </ul>;
  }

  render() { return (
    <TabMenu className="filter-menu">
      <Tab active name="All" render={this.showAllTodos} />
      <Tab name="Incomplete" render={this.showIncompleteTodos} />
      <Tab name="Completed" render={this.showCompletedTodos} />
    </TabMenu>
  );}
}

TodoList.propTypes = {
  todos: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string.isRequired,
    completed: PropTypes.bool
  })),
  onUpdateTodo: PropTypes.func.isRequired,
  toggleTodo: PropTypes.func,
};

TodoList.defaultProps = {
  todos: []
};

const mapStateToProps = ({ todos }) => ({ todos: TodoModel.wrap(todos) });
export default connect(mapStateToProps, { toggleTodo })(TodoList);
