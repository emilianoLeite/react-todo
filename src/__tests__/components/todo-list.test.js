import React from "react";
import { TodoList } from "../../components";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

describe("when no todos are supplied", () => {
  it("renders a empty list", () => {
    const wrapper = shallow(<TodoList onClickedTodo={() => { }} />);

    expect(wrapper.find("ul").children()).toHaveLength(0);
  });
});

describe("when some todos are supplied", () => {
  it("matches snapshot", () => {
    const todos = [
      { text: "todo1" },
      { text: "todo2", completed: true },
    ];
    const tree = renderer
      .create(<TodoList onClickedTodo={() => { }} todos={todos} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe("when a todo's checkbox is clicked", () => {
    it("emits the index of the clicked todo to the supplied callback", () => {
      const handleCompletedTodoCallback = jest.fn();
      const todos = [{text: "Todo", completed: false}];
      const wrapper = shallow(<TodoList
        onClickedTodo={handleCompletedTodoCallback}
        todos={todos}
      />);

      wrapper.find("input[type='checkbox']").simulate("click");

      expect(handleCompletedTodoCallback).toHaveBeenCalledWith(0);
    });
  });
});
