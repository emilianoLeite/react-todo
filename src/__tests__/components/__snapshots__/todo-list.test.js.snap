// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when some todos are supplied matches snapshot 1`] = `
Array [
  <div
    className="filter-menu"
  >
    <div
      className="menu-item active"
      onClick={[Function]}
    >
      All
    </div>
    <div
      className="menu-item"
      onClick={[Function]}
    >
      Incomplete
    </div>
    <div
      className="menu-item"
      onClick={[Function]}
    >
      Completed
    </div>
  </div>,
  <ul
    className="todo-list"
  >
    <li
      className=""
    >
      <input
        onClick={[Function]}
        type="checkbox"
      />
      <span
        className="todo-text"
        onClick={[Function]}
      >
        todo1
      </span>
    </li>
    <li
      className="completed"
    >
      <input
        defaultChecked={true}
        onClick={[Function]}
        type="checkbox"
      />
      <span
        className="todo-text"
        onClick={[Function]}
      >
        todo2
      </span>
    </li>
  </ul>,
]
`;
